Execution of Windows executables broken in 1.5.0

I'm using exec-maven-plugin on Windows to run `npm install` in a JS project. 
```
<groupId>org.codehaus.mojo</groupId>
<artifactId>exec-maven-plugin</artifactId>
<version>1.4.0</version>
<executions>
	<execution>
		<id>npm install</id>
		<phase>generate-resources</phase>
		<goals>
			<goal>exec</goal>
		</goals>
		<configuration>
			<executable>npm</executable>
			<commandlineArgs>install --registry ${npm.registry}</commandlineArgs>
		</configuration>
	</execution>
</executions>
```

which works fine with 1.4.0 but switching to 1.5.0 results in
```
[INFO] --- exec-maven-plugin:1.5.0:exec (npm install) @ util ---
[ERROR] Command execution failed.
java.io.IOException: Cannot run program "C:\Program Files\nodejs\npm" (in directory "D:\Workspaces\git-repositories\myproject"): CreateProcess error=193, %1 ist keine zulässige Win32-Anwendung
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
        at java.lang.Runtime.exec(Runtime.java:620)
        at org.apache.commons.exec.launcher.Java13CommandLauncher.exec(Java13CommandLauncher.java:61)
        at org.apache.commons.exec.DefaultExecutor.launch(DefaultExecutor.java:279)
        at org.apache.commons.exec.DefaultExecutor.executeInternal(DefaultExecutor.java:336)
        at org.apache.commons.exec.DefaultExecutor.execute(DefaultExecutor.java:166)
        at org.codehaus.mojo.exec.ExecMojo.executeCommandLine(ExecMojo.java:764)
        at org.codehaus.mojo.exec.ExecMojo.executeCommandLine(ExecMojo.java:711)
        at org.codehaus.mojo.exec.ExecMojo.execute(ExecMojo.java:289)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: java.io.IOException: CreateProcess error=193, %1 ist keine zulässige Win32-Anwendung
        at java.lang.ProcessImpl.create(Native Method)
        at java.lang.ProcessImpl.<init>(ProcessImpl.java:386)
        at java.lang.ProcessImpl.start(ProcessImpl.java:137)
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
        ... 30 more
```

--- a/src/main/java/org/codehaus/mojo/exec/ExecMojo.java
+++ b/src/main/java/org/codehaus/mojo/exec/ExecMojo.java
@@ -663,15 +663,14 @@ public class ExecMojo
         search: for ( final String path : paths )
         {
             f = new File( path, executable );
+            for ( final String extension : getExecutableExtensions() )
+            {
+                f = new File( path, executable + extension );
+                if ( f.isFile() )
+                    break search;
+            }
             if ( f.isFile() )
-                break;
-            else
-                for ( final String extension : getExecutableExtensions() )
-                {
-                    f = new File( path, executable + extension );
-                    if ( f.isFile() )
-                        break search;
-                }
+              break;
         }
 
         if ( f == null || !f.exists() )

--- a/src/main/java/org/codehaus/mojo/exec/ExecMojo.java
+++ b/src/main/java/org/codehaus/mojo/exec/ExecMojo.java
@@ -663,14 +663,15 @@ public class ExecMojo
         search: for ( final String path : paths )
         {
             f = new File( path, executable );
-            for ( final String extension : getExecutableExtensions() )
-            {
-                f = new File( path, executable + extension );
-                if ( f.isFile() )
-                    break search;
-            }
-            if ( f.isFile() )
-              break;
+            if ( !OS.isFamilyWindows() && f.isFile() )
+                break;
+            else
+                for ( final String extension : getExecutableExtensions() )
+                {
+                    f = new File( path, executable + extension );
+                    if ( f.isFile() )
+                        break search;
+                }
         }
 
         if ( f == null || !f.exists() )

--- a/src/test/java/org/codehaus/mojo/exec/ExecMojoTest.java
+++ b/src/test/java/org/codehaus/mojo/exec/ExecMojoTest.java
@@ -317,7 +317,10 @@ public class ExecMojoTest
     public void testGetExecutablePathPreferExecutableExtensionsOnWindows()
     		throws IOException
     {
-    	Assume.assumeTrue(OS.isFamilyWindows());
+    	// this test is for Windows
+    	if (!OS.isFamilyWindows()) {
+    		return;
+    	}
     	final ExecMojo realMojo = new ExecMojo();
     	
     	final String tmp = System.getProperty("java.io.tmpdir");

--- a/src/test/java/org/codehaus/mojo/exec/ExecMojoTest.java
+++ b/src/test/java/org/codehaus/mojo/exec/ExecMojoTest.java
@@ -35,7 +35,11 @@ import org.apache.maven.project.MavenProject;
 import org.apache.maven.project.MavenProjectBuilder;
 import org.codehaus.plexus.logging.Logger;
 import org.codehaus.plexus.logging.console.ConsoleLogger;
+import org.codehaus.plexus.util.IOUtil;
 import org.codehaus.plexus.util.StringOutputStream;
+import org.junit.Assume;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
 
 /**
  * @author Jerome Lacoste <jerome@coffeebreaks.org>
@@ -61,7 +65,7 @@ public class ExecMojoTest
 
         public String failureMsg;
 
-        public Map systemProperties = new HashMap();
+        public Map<String, String> systemProperties = new HashMap<String, String>();
 
         protected int executeCommandLine( Executor exec, CommandLine commandLine, Map enviro, OutputStream out,
                                           OutputStream err )
@@ -77,7 +81,7 @@ public class ExecMojoTest
 
         protected String getSystemProperty( String key )
         {
-            return (String) systemProperties.get( key );
+            return systemProperties.get( key );
         }
 
         int getAmountExecutedCommandLines()
@@ -87,7 +91,7 @@ public class ExecMojoTest
 
         CommandLine getExecutedCommandline( int index )
         {
-            return ( (CommandLine) commandLines.get( index ) );
+            return commandLines.get( index );
         }
     }
 
@@ -268,7 +272,7 @@ public class ExecMojoTest
         ExecMojo realMojo = new ExecMojo();
 
         File workdir = new File( System.getProperty( "user.dir" ) );
-        Map enviro = new HashMap();
+        Map<String, String> enviro = new HashMap<String, String>();
 
         String myJavaPath = "target" + File.separator + "javax";
         File f = new File( myJavaPath );
@@ -305,6 +309,42 @@ public class ExecMojoTest
             f.delete();
             assertFalse( "file deleted...", f.exists() );
         }
+    }
+    
+    // under Windows cmd/bat files should be preferred over files with the same prefix without extension, e.g.
+    // if there are "node" and "node.cmd" the mojo should pick "node.cmd"
+    // see https://github.com/mojohaus/exec-maven-plugin/issues/42
+    public void testGetExecutablePathPreferExecutableExtensionsOnWindows()
+    		throws IOException
+    {
+    	Assume.assumeTrue(OS.isFamilyWindows());
+    	final ExecMojo realMojo = new ExecMojo();
+    	
+    	final String tmp = System.getProperty("java.io.tmpdir");
+    	final File workdir = new File(tmp, "testGetExecutablePathPreferExecutableExtensionsOnWindows");
+    	workdir.mkdirs();
+    	
+    	final Map<String, String> enviro = new HashMap<String, String>();
+    	
+    	final File f = new File(workdir, "mycmd");
+    	final File fCmd = new File(workdir, "mycmd.cmd");
+    	f.createNewFile();
+    	fCmd.createNewFile();
+    	assertTrue( "file exists...", f.exists() );
+    	assertTrue( "file exists...", fCmd.exists() );
+    	
+    	realMojo.setExecutable( "mycmd" );
+    	
+    	final CommandLine cmd = realMojo.getExecutablePath( enviro, workdir );
+    	// cmdline argumets are: [/c, %path-to-temp%\mycmd.cmd], so check second argument
+    	assertTrue( "File should have cmd extension",
+    			cmd.getArguments()[1].endsWith( "mycmd.cmd" ) );
+    	
+    	f.delete();
+    	fCmd.delete();
+    	assertFalse( "file deleted...", f.exists() );
+    	assertFalse( "file deleted...", fCmd.exists() );
+    	
     }
 
     public void testRunFailure()

